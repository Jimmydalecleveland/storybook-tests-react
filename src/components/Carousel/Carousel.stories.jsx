import Carousel from "."

const Story = {
  title: 'Example/Carousel',
  component: Carousel
}

const Template = (args) => <Carousel {...args} />

export const Default = Template.bind({})
Default.args = {
  slides: ["Swashbuckling", "with", "Code"]
}

export default Story