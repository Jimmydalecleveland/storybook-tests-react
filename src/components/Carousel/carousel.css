.carousel-container {
  height: 25vw;
  width: 45vw;
  display: grid;
  grid-template: 1 / 1;
  font-size: 6rem;
}

.carousel-slide {
  grid-area: 1 / 1;
  transition: all 0.3s 0.3s cubic-bezier(0.5, 0, 0.5, 1);
  transition-property: opacity, transform;

  display: block;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 0.5rem;
  box-shadow: 0 1rem 2rem rgba(black, 0.1);
}

.carousel-slide[hidden] {
  transition-delay: 0s;
  opacity: 0;
  transform: scale(0.8);
}

.carousel-slide:nth-child(1) {
  background: linear-gradient(45deg, #0e192c, #21375c);
  color: #fa9474;
}

.carousel-slide:nth-child(2) {
  background: linear-gradient(45deg, #0e192c, #21375c);
  color: #67ddff;
}

.carousel-slide:nth-child(3) {
  background: linear-gradient(45deg, #0e192c, #21375c);
  color: #40ffdc;
}
