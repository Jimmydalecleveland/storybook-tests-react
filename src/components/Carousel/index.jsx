// Based on David Khourshid's codesandbox example:
// https://codesandbox.io/s/funny-wind-p59ei?file=/src/index.js
import { useState } from "react";
import PropTypes from "prop-types";
import "./carousel.css";

function Carousel({ slides }) {
  const [index, setIndex] = useState(0);

  return (
    <section
      data-testid="carousel-container"
      className="carousel-container"
      onClick={() => setIndex((index + 1) % slides.length)}
    >
      {slides.map((screen, i) => (
        <article
          className="carousel-slide"
          key={i}
          hidden={i !== index || undefined}
        >
          {screen}
        </article>
      ))}
    </section>
  );
}

Carousel.propTypes = {
  slides: PropTypes.array.isRequired,
};

export default Carousel;
