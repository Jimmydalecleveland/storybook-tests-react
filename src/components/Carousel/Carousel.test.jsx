import { Default } from "./Carousel.stories";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

test("should render", () => {
  const { container } = render(<Default {...Default.args} />);
  expect(container.firstChild).toMatchSnapshot();
});

test("should switch slides when clicked", () => {
  render(<Default {...Default.args} />);
  const carouselContainer = screen.getByTestId("carousel-container");

  // Grab all slides, as they are always in the DOM
  const firstSlide = screen.getByText(/swashbuckling/i);
  const secondSlide = screen.getByText(/with/i);
  const thirdSlide = screen.getByText(/code/i);

  // First slide is visible by default
  expect(firstSlide).toBeVisible();
  expect(secondSlide).not.toBeVisible();
  expect(thirdSlide).not.toBeVisible();

  // After clicking once, only the Second slide is visible
  userEvent.click(carouselContainer);
  expect(firstSlide).not.toBeVisible();
  expect(secondSlide).toBeVisible();
  expect(thirdSlide).not.toBeVisible();
});
