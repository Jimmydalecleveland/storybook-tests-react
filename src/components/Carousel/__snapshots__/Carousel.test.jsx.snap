// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<section
  class="carousel-container"
  data-testid="carousel-container"
>
  <article
    class="carousel-slide"
  >
    Swashbuckling
  </article>
  <article
    class="carousel-slide"
    hidden=""
  >
    with
  </article>
  <article
    class="carousel-slide"
    hidden=""
  >
    Code
  </article>
</section>
`;
